File: structures/Bin_heap.java
import java.io.*;
import java.util.*;

public class Bin_heap {
    public static void main(String[] args) throws FileNotFoundException {

        int n;
        int[] arr;

        try (Scanner sc = new Scanner(new FileReader("input.txt"))) {
            n = sc.nextInt();
            arr = new int[n + 1];
            int i = 1;
            while (sc.hasNext()) {
                arr[i] = sc.nextInt();
                i++;
            }
        }

        try (PrintWriter pw = new PrintWriter(new File("output.txt"))){
            for (int i = n; i >= 2;) {
                if (arr[i] >= arr[i / 2]) 
                i--;
                else {
                    pw.print("No");
                    pw.close();
                    System.exit(0);
                }
            }
            pw.print("Yes");
            pw.close();
        }
    }
}

File: structures/Bin_heap.cpp
#include <iostream>
#include <cmath>
#include <fstream>

using namespace std;

int main() {
    ifstream is("input.txt");
    ofstream os("output.txt");

    long long n = 0;
    is >> n;

    int len = (int) log2(n) + 1;
    bool* res = new bool[len];
    int i = 0;

    while(n != 0) {
        res[i++] = n % 2;
        n /= 2;
    }

    for(int i = 0; i < len; i++) if(res[i]) os << i << "\n";
}